### DISCUSSION TITLE
Should it be possible to use `Unpack` in `Union`?

### DISCUSSION THREAD

**Discussion by sobolevn** [python/typing#1824] [Q&A]
I have the following function:

```python
def random_tuple_element(t: tuple):
     return random.choice(t)
```

My tuples can have any structure. `[int, int, int]` or `[str, int, bool, CustomObject]` or whatever.

I thought that something like this should be possible:

```python
import random
from typing import Union
from typing_extensions import TypeVarTuple

Ts = TypeVarTuple('Ts')

def random_tuple_element(t: tuple[*Ts]) -> Union[*Ts]:
     return random.choice(t)
```

But, right now it is not possible: https://mypy-play.net/?mypy=latest&python=3.12&gist=5ee956588f2eb33faba76e87902cf59a

```python
# E: Unpack is only valid in a variadic position
```

Does it make sense to allow `Unpack` in `Union`?

**Comment by erictraut** [comment]
Unpacking a TypeVarTuple within a `Union` was supported in earlier drafts of [PEP 646](https://peps.python.org/pep-0646/), but it was [removed](https://github.com/python/peps/pull/1856) before the final draft was submitted to the SC for approval.

I implemented support for this in pyright (based on the earlier draft PEP). It was quite difficult to add, so I empathize with the PEP authors who justified the removal based on a lack of use case plus feedback from the pyre team. I'll note that pyre and pytype still don't have support for TypeVarTuple in general, and it has been four years since PEP 646 was finalized.

There have been some requests to add back this capability. See [here](https://discuss.python.org/t/proposal-allow-typevartuple-unpacking-in-unions/42932) and [here](https://github.com/python/typing/issues/1523). None of these have progressed beyond an initial discussion.

Pyright still contains my original implementation (I didn't have the heart to rip it out completely after spending so much time getting it working), but it's enabled only if you set `enableExperimentalFeatures` to true. Here's an example in the [pyright playground](https://pyright-play.net/?reportMissingModuleSource=true&enableExperimentalFeatures=true&code=JYWwDg9gTgLgBFAhgOwCYRAKAGZQ3GATzGGQHM5RJY4BVZYCZAGjgBViBTANUSjYCuYADadMmNgGc4AXnZde-IaIAUAcilqAlONSdsCFOhAB9GMs4nOokJ2QwVMAFwELAbQBUUgLpa4AWgA%2BOgYmTx8nTDho6KhOcyhkQzQMADoAYwALCGB0zkcdTD0DbAFkdJVCF3MRTjdSGFZJGChWDxrRN2FgZvr7b1ZUoe9fSJi4AA9ZZOMzCysbOwdCHXG4gDdORGEzLhUJvwBiSns4AB84ZqhzuG7ehu9MIA).

If you'd like to see this added to the typing spec, you could propose an update to the spec. This change is probably sufficiently small that it wouldn't require a full PEP, but there should be a discussion with an opportunity for type checker maintainers to weigh in.

**Comment by AstraLuma** [comment]
Use case: type describing JSON, extensibly.

```python
type JSONish[*T] = Union[str, int, float, bool, None, dict[str, Jsonish[*T]], list[Jsonish[*T]], *T]

class Foo: ...

strict: JSONish
looser: JSONish[Foo]
```

**Reply by AstraLuma** [reply]
Of course, you could write:

```python
type JSONish[T] = Union[str, int, float, bool, None, dict[str, Jsonish[T]], list[Jsonish[T]], T]

class Foo: ...

strict: JSONish
looser: JSONish[Foo|object]
```
